pkg/topology/full/full.go

```diff
@@ -20,81 +21,111 @@ func init() {
+		discovery:     disc,
+		addressBook:   addressBook,
+		connecter:     connecter,
+		receivedPeers: make(map[string]bool),
+		logger:        logger,
 	}
 }
 
 // AddPeer adds a new peer to the topology driver.
 // The peer would be subsequently broadcasted to all connected peers.
 // All conneceted peers are also broadcasted to the new peer.
-func (d *driver) AddPeer(overlay swarm.Address) error {
-	d.mtx.Lock()
-	defer d.mtx.Unlock()
+func (d *Driver) AddPeer(addr swarm.Address) error {
+	if d.receivedPeers[addr.ByteString()] {
+		return nil
+	}
 
-	ma, exists := d.addressBook.Get(overlay)
+	d.receivedPeers[addr.ByteString()] = true
+	connectedPeers := d.connecter.Peers()
+	ma, exists := d.addressBook.Get(addr)
 	if !exists {
 		return topology.ErrNotFound
 	}
 
-	var connectedNodes []discovery.BroadcastRecord
-	for _, addressee := range d.connected {
-		cma, exists := d.addressBook.Get(addressee)
-		if !exists {
-			return topology.ErrNotFound
-		}
-
-		err := d.discovery.BroadcastPeers(context.Background(), addressee, discovery.BroadcastRecord{Overlay: overlay, Addr: ma})
+	if !isConnected(addr, connectedPeers) {
+		peerAddr, err := d.connecter.Connect(context.Background(), ma)
 		if err != nil {
 			return err
 		}
 
-		connectedNodes = append(connectedNodes, discovery.BroadcastRecord{Overlay: addressee, Addr: cma})
+		// update addr if it is wrong or it has been changed
+		if !addr.Equal(peerAddr) {
+			addr = peerAddr
+			d.addressBook.Put(peerAddr, ma)
+		}
+	}
+
+	connectedAddrs := []swarm.Address{}
+	for _, addressee := range connectedPeers {
+		// skip curenttly added peer
+		if addressee.Address.Equal(addr) {
+			continue
+		}
+
+		connectedAddrs = append(connectedAddrs, addressee.Address)
+		if err := d.discovery.BroadcastPeers(context.Background(), addressee.Address, addr); err != nil {
+			return err
+		}
 	}
 
-	err := d.discovery.BroadcastPeers(context.Background(), overlay, connectedNodes...)
-	if err != nil {
+	if err := d.discovery.BroadcastPeers(context.Background(), addr, connectedAddrs...); err != nil {
 		return err
 	}
 
-	// add new node to connected nodes to avoid double broadcasts
-	d.connected[overlay.String()] = overlay
 	return nil
 }
 
 // ChunkPeer is used to suggest a peer to ask a certain chunk from.
-func (d *driver) ChunkPeer(addr swarm.Address) (peerAddr swarm.Address, err error) {
-	d.mtx.Lock()
-	defer d.mtx.Unlock()
-
-	if len(d.connected) == 0 {
+func (d *Driver) ChunkPeer(addr swarm.Address) (peerAddr swarm.Address, err error) {
+	connectedPeers := d.connecter.Peers()
+	if len(connectedPeers) == 0 {
 		return swarm.Address{}, topology.ErrNotFound
 	}
 
-	itemIdx := rand.Intn(len(d.connected))
+	itemIdx := rand.Intn(len(connectedPeers))
 	i := 0
-	for _, v := range d.connected {
+	for _, v := range connectedPeers {
 		if i == itemIdx {
-			return v, nil
+			return v.Address, nil
 		}
 		i++
 	}
 
 	return swarm.Address{}, topology.ErrNotFound
 }
+
+func (d *Driver) PeerAddedHandler(addr swarm.Address) {
```

Comment:
Why do we need this?

Reply:
To avoid calling topology driver directly from hive. It's what we talked about a circular dependency, handelrs are one way to acomplish this. Right?

Reply:
Take a look at the node to see how all the parts (hive, topology driver etc..) are initialized and connected.

Reply:
oops sorry is missed that.
do we have to have this in a separate goroutine though? is this not premature? or does this break tests? deadlocks? please elaborate

Reply:
i wonder if this is actually needed as an explicit method tbh.
you could just inject the handler in the when registering like so:
```go
	p2ps.SetPeerAddedHandler(func(addr swarm.Address){ topologyDriver.AddPeer(addr) })
```

Reply:
or even nicer, change the signatures that you can use `AddPeer` directly without having to wrap it

Reply:
Hmmm I have 2 points about this, let me know what you think:
1. I think it is a bad practice to block the handler of a different service
2. Doing it as a function in node makes node know about business logic, which I also don't like

What is a problem with having a special goroutine for a simple implementation, and a special function? Any downsides?

Reply:
The signature is not the same, but similar just because I don't actually want to handle error from handlers in hive, I think it adds no value, but that can be done as well. My goal was to separate this.

Reply:
Well, if you adjust the signature of the function that you're expecting to be `func(swarm.Address) error` and then just directly use `AddPeer` in `SetPeerAddedHandler` (but then the `SetHandler` naming should probably also change).

About goroutines, well, it introduces concurrency which is not necessarily trivial and could lead to flaky tests and certain data races, especially when the components work with the underlying data structures, so obviously we should make sure that the necessary changes are made to the address book first for example, but this is not so clear from the actual code (and someone else tampering with this code might not see this at first glance), and when we add more indirection and more concurrency this becomes less and less clear

Reply:
@acud , wrapping one line of code was done because our linter complains if function is just called with `go` and error is not handled. I personally don't like it either :)

Reply:
And in the more complex implementation (see todo comment) we could add various things here.

Reply:
I think that it is ok to have a todo comment about what it needs to be improved in the next iteration. And yes, the linter warning is handled making it obvious that the error and goroutine is yet to be handled. It is still needed to be tested where it should be improved.

Reply:
the questions was also if we need this invoked in a different goroutine @janos

Reply:
Just my 2 c, making handler blocking could also lead to additional problem in tests, because hive handler that receives messages is also async by nature. And I would not try to solve the problem for long blocking handlers, but for async handlers cuz it looks to me that this is what we would need.

Reply:
Since this is a first implementation, I don't think it makes sense to argue too much about it now, and both solution are good nenouigh, so I will switch to a blocking implementation and your suggestion @acud

Reply:
Yes, let's try the blocking implementation and have it async if we experience problems.

Reply:
thanks guys